### ISSUE TITLE
Mouse scroll misbehaves on Linux after alt-tabbing

### ISSUE THREAD

**Issue by jaromero** [atom/atom#12591]
### Prerequisites

* [x] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode)?
* [x] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?
* [x] Did you check the [debugging guide](http://flight-manual.atom.io/hacking-atom/sections/debugging/)?
* [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
* [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
* [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

For more information on how to write a good [bug report](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-bug-report) or [enhancement request](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-enhancement-suggestion), see the `CONTRIBUTING` guide.

### Description

With a window that partially or fully overlaps Atom's window, if that other window is on top and mouse scrolls are made on the part that overlaps Atom, and then one alt-tabs directly back into Atom and attempts to mouse scroll, it "re-plays" all those scrolls made and applies them along with the mouse scroll actually made.

### Steps to Reproduce

1. Open Atom, preferably with a very long file, open around the middle, to make it more obvious
2. Open any other window and place it on top of Atom. It doesn't actually matter if it has any scrollable content at all.
3. Focus the other window and mouse scroll a lot over it in the same direction. Say, downwards.
4. Alt-tab to Atom. Make sure not to move your mouse, at least not outside the boundaries of the Atom window.
5. Perform a single mouse scroll "click" in the opposite direction, i.e. if you scrolled downwards a lot in step 3, now scroll upwards once
6. Repeat steps 3 to 5, swap directions around a bit

**Expected behavior:** The content scrolls up and down accordingly, only as far as I expect it to

**Actual behavior:** The first time you do step 6, no scrolls happen. Afterwards, the content scrolls by as much as you've scrolled in step 3, and in that direction, even though in step 6 you scrolled once.

[I've recorded a video demonstrating this](https://my.mixtape.moe/nrpgkw.webm). Note that each time I alt-tab to Atom, I do a single scroll 'click', but the content jumps by a lot each time.

Switching to Atom by any method other than alt-tab (i.e. clicking on the Atom window to raise it, clicking the icon on any dock or taskbar, etc.) does not trigger this behaviour. Only alt-tabbing, and only if the mouse cursor doesn't leave the boundaries of the Atom window right before or after alt-tabbing.

Atom 1.9 also had the same problem, which I worked around by downgrading electron to 0.36.12 at compile time. This workaround makes Atom 1.10 fail to start (unrelated issue), so downgrading electron is no longer an option for me.

[This Chrome issue](https://bugs.chromium.org/p/chromium/issues/detail?id=608246) that I've also reported is probably related, but now it happens even without xf86-xinput-libinput or even libinput installed.

### Versions

~~~
Atom    : 1.10.0
Electron: 0.37.8
Chrome  : 49.0.2623.75
Node    : 5.10.0
~~~

Atom installed from this [AUR package](https://aur.archlinux.org/packages/atom-editor/)
OS: Arch Linux 64-bit, fully up-to-date.

**Comment by andyrichardson** [comment]
I tried your suggestions and I failed to reproduce the error. Scroll events in one window/application should not affect another, this must be an issue with your window manager/desktop environment and not Atom itself.

---

The causes of this issue are likely due to a beta/unstable version of _Gnome shell_ or the use of a non X display server such as Wayland.

**My System**
Fedora 24
GNOME Shell 3.20.4
X.Org X Server 1.18.4

**Comment by jaromero** [comment]
I'm running X 1.18.4 and Gnome Shell 3.20.4 as well, I should've put that in my OP.

But I've also just tested it in a fresh (but updated) Xubuntu 16.04 VM and the issue is still present for me.

**Comment by andyrichardson** [comment]
I'm not sure what to suggest but I think the issue here is with Atom/Electron receiving a scroll event from the system and therefore not an issue with Atom itself - I can't be 100% sure about that, however.

Do you have another input device to test this out on? Also, do you have any smooth scrolling functionality enabled?

**Comment by jaromero** [comment]
It might be, but no other app I've seen so far shows this issue, not even Chrome. It definitely did not happen with Atom 1.9 (and electron downgraded).

I've tested with my laptop's trackpad, as well as an external Logitech T650. All show the same issue.

I'll see if I can get a different laptop to test on. And test here with a VM under Windows, just to rule out it being a local OS issue that is somehow propagating into my VMs.

**Comment by jaromero** [comment]
I forgot to mention that I don't have any smooth scrolling enabled anywhere.

I just spun up another fresh xubuntu VM (under Windows this time) and recorded the results with both Atom 1.9.9 and 1.10. The mouse cursor isn't visible, but I kept it within the bounds of the terminal window during both tests:

https://my.mixtape.moe/wyxtuh.webm
https://my.mixtape.moe/abvelo.webm

Both are, again, fresh xubuntu VMs, fully updated, no additional repositories or tweaks done, no other packages installed other than those required by Atom itself.

**Comment by andyrichardson** [comment]
Hmmm interesting! I'm not managing to replicate the issue but there's a good chance that It's an issue with Electron, based on your discoveries. I'd advise you try this with another electron app - [Nylas](https://www.nylas.com/), [Slack](https://slack.com/) or [any of these](http://electron.atom.io/#apps).

If you're able to replicate this with another app and with the latest version of Electron, we can look into pushing this issue to the [Electron repo](https://github.com/electron/electron) as well as keeping this one open.

**Comment by Tribex** [comment]
I've been able to replicate this, and it's been rather annoying of late: Ubuntu GNOME 16.04, GNOME 3.20, Atom 1.10 & 1.11.

**Comment by jaromero** [comment]
Testing with a few other electron apps.

The following show this exact same issue. Electron version stated where known:

## [VS Code](https://code.visualstudio.com/)

**Version:** 1.4.0
**Electron:** 0.37.6

## [Boostnote](https://b00st.io/)

**Version:** 0.6.1
**Electron:** [1.2.8](https://github.com/BoostIO/Boostnote/blob/master/package.json#L16)

-----

The following apps **do not** show this issue:

## [Brave](https://github.com/brave/browser-laptop/blob/master/docs/linuxInstall.md)

**Version:** 0.11.6
**Electron:** 1.3.8

## [SimpleNote](https://simplenote.com/)

**Version:** 1.0.3
**Electron:** [1.3.2 I think?](https://github.com/Automattic/simplenote-electron/blob/master/package.json#L32)

## [Marp](https://yhatt.github.io/marp/)

**Version:** 0.0.8
**Electron:** [1.2.7](https://github.com/yhatt/marp/blob/master/package.json#L21)

## [HyperTerm](https://github.com/zeit/hyperterm)

**Version:** (git master)
**Electron:** 1.3.5

## [GitKraken](https://www.gitkraken.com/)

**Version:** 1.5.4
**Electron:** Unknown

## [LightTable](http://lighttable.com/)

**Version:** 0.8.1
**Electron:** 0.34.5

## [Abricotine](http://abricotine.brrd.fr/)

**Version:** 0.3.3
**Electron:** [0.35.0](https://github.com/brrd/Abricotine/blob/master/package.json#L35)

## [KeeWeb](https://keeweb.info/)

**Version:** 1.3.1
**Electron:** [1.3.3](https://github.com/keeweb/keeweb/blob/master/package.json#L18)

## [SQLectron-GUI](https://sqlectron.github.io/)

**Version:** 1.13
**Electron:** [1.2.7](https://github.com/sqlectron/sqlectron-gui/blob/master/package.json#L77)

-----

That's about an hour of testing, I basically went down the list of apps.

The ones that don't work properly seem to be using electron 0.37.x (Atom, VSCode) or 1.2.8 (Boostnote). But those that I saw were using 1.2.7 were working just fine. Apps using 1.3.x or higher seem to work fine.

So it looks like sometime after 0.37.x and before 1.2.7 either the issue was resolved or devs worked around it in _some_ way, except for Boostnote which is still broken.

I'll try and get a minimal sample electron app to reproduce this, but I agree that it does seem to be in the end an electron issue. Atom just happens to be a very large portion of my workflow as a web dev.

**Comment by andyrichardson** [comment]
Great work! It's rather absurd that Atom is using such an outdated version of Electron given the fact that  Electron was built with Atom in mind.

---


To conclude this issue, the solution is for Atom to update Electron to v1.2.7 or later

@50Wliu I think sufficient evidence has been provided to justify removing the 'needs-reproduction' label. Also, is it possible to now label this as a known issue or is more reproduction required?

**Comment by 50Wliu** [comment]
Will be resolved by #12300.
